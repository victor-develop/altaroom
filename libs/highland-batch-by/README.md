# highland-batch-by

This library was generated with [Nx](https://nx.dev).

## Building

Run `nx build highland-batch-by` to build the library.

## Running unit tests

Run `nx test highland-batch-by` to execute the unit tests via [Jest](https://jestjs.io).
