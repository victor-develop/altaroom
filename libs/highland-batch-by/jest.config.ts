/* eslint-disable */
export default {
  displayName: 'highland-batch-by',
  preset: '../../jest.preset.js',
  testEnvironment: 'node',
  transform: {
    '^.+\\.[tj]s$': ['ts-jest', { tsconfig: '<rootDir>/tsconfig.spec.json' }],
  },
  moduleFileExtensions: ['ts', 'js', 'html'],
  coverageDirectory: '../../coverage/libs/highland-batch-by',
};
