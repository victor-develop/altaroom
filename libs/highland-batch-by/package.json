{
  "name": "@altaroom/highland-batch-by",
  "version": "0.0.1",
  "dependencies": {
    "highland": "^2.13.5"
  },
  "type": "commonjs",
  "main": "./index.cjs",
  "devDependencies": {
    "@types/highland": "^2.12.16"
  }
}
