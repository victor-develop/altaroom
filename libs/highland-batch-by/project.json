{
  "name": "highland-batch-by",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/highland-batch-by/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nx/esbuild:esbuild",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/libs/highland-batch-by",
        "main": "libs/highland-batch-by/src/index.ts",
        "tsConfig": "libs/highland-batch-by/tsconfig.lib.json",
        "assets": ["libs/highland-batch-by/*.md"],
        "generatePackageJson": true,
        "format": ["cjs"]
      }
    },
    "publish": {
      "command": "node tools/scripts/publish.mjs highland-batch-by {args.ver} {args.tag}",
      "dependsOn": ["build"]
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": [
          "libs/highland-batch-by/**/*.ts",
          "libs/highland-batch-by/package.json"
        ]
      }
    },
    "test": {
      "executor": "@nx/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/highland-batch-by/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  },
  "tags": []
}
