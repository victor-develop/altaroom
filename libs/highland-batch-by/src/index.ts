export * from './lib/highland-batch-by';
