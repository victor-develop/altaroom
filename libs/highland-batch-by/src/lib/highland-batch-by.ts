import * as highland from 'highland'

export interface Control<TBatch, TEntry, TPushedBatch> {
  beginWith: () => TBatch;
  forEachEntry: (b: TBatch, e: TEntry) => { batch_to_update: TBatch; batch_to_push: TPushedBatch | null };
  end: (last_batch: TBatch) => { batch_to_push: TPushedBatch | null };
}

export function batchBy<TBatch, TEntry, TPushedBatch>(control: Control<TBatch, TEntry, TPushedBatch>): (object_stream: Highland.Stream<TEntry>) => Highland.Stream<TPushedBatch> {
  return function(object_stream: Highland.Stream<TEntry>): Highland.Stream<TPushedBatch> {
    const output_stream: Highland.Stream<TPushedBatch> = highland();
    object_stream.scan(control.beginWith(), (batch, entry) => {
      const result = control.forEachEntry(batch, entry);
      if (result.batch_to_push) {
        output_stream.write(result.batch_to_push);
      }
      return result.batch_to_update;
    })
      .last()
      .map(last_batch => {
        const result = control.end(last_batch);
        result.batch_to_push && output_stream.write(result.batch_to_push);
      })
      .done(() => {
        output_stream.end()
      });
      
    return output_stream;
  }
}


export interface TBatchSamePattern<TPattern, TEntry> {
  accumulate_pattern: TPattern | null;
  items: TEntry[];
}

const samePattern = <TEntry, TPattern>({
  extractPattern,
  same,
}: {
  extractPattern: (entry: TEntry, accumulate_pattern: TPattern | null) => TPattern;
  same: (a: TPattern, b: TPattern) => boolean;
}): Control<TBatchSamePattern<TPattern, TEntry>, TEntry, TBatchSamePattern<TPattern, TEntry> | null> => {
  return {
    beginWith: () => ({ accumulate_pattern: null, items: [] }),
    forEachEntry: (batch, entry) => {
      const { accumulate_pattern, items } = batch;
      const current_pattern = extractPattern(entry, accumulate_pattern);
      if (accumulate_pattern === null || same(accumulate_pattern, current_pattern)) {
        return {
          batch_to_push: null,
          batch_to_update: {
            accumulate_pattern: current_pattern,
            items: [...items, entry],
          },
        };
      } else {
        return {
          batch_to_push: {
            accumulate_pattern,
            items,
          },
          batch_to_update: {
            accumulate_pattern: current_pattern,
            items: [entry],
          },
        };
      }
    },
    end: (last_batch) => {
      return {
        batch_to_push: {
          accumulate_pattern: last_batch.accumulate_pattern,
          items: last_batch.items,
        },
      };
    },
  };
};

const sameProperty = <TEntry, TKey extends keyof TEntry>(
  propName: TKey
) => {
  return samePattern<TEntry, TEntry[TKey]>({
    extractPattern: x => x[propName],
    same: (p1, p2) => p1 === p2
  });
};

/**
 * 
 * @example
 * Example using it with highland.through
 * 
 * ```
 * highland([
 *   {sku:'a', order_id:1},
 *   {sku:'b', order_id:1},
 *   {sku:'c', order_id:2}
 * ]).through(batchByProp('order_id'))
 * .each(console.log)
 * .done(() => {})
 * ```
 * Output:
 * 
 * ```
 * [
 *   {accumulate_pattern: 1, items: [{sku:'a', order_id:1}, {sku:'b', order_id:1}]},
 *   {accumulate_pattern: 2, items: [{sku:'c', order_id:2}]}
 * ]
 * ```
 * 
 * @param propName 
 * @returns 
 */
export const batchByProp = <TEntry, TKey extends keyof TEntry>(propName: TKey) => batchBy(sameProperty<TEntry, TKey>(propName))

