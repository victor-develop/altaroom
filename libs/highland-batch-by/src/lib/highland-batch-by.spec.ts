import { batchByProp } from './highland-batch-by';
import * as highland from 'highland'

describe('batchByProp', () => {
  it('should work', async () => {
    const x = await highland([
       {sku:'a', order_id:1},
       {sku:'b', order_id:1},
       {sku:'c', order_id:2}
    ])
    .through(batchByProp('order_id'))
    .collect()
    .toPromise(Promise)
    expect(x).toMatchSnapshot('output')
  });
});
