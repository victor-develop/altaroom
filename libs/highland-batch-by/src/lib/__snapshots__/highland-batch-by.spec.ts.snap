// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`batchByProp should work: output 1`] = `
[
  {
    "accumulate_pattern": 1,
    "items": [
      {
        "order_id": 1,
        "sku": "a",
      },
      {
        "order_id": 1,
        "sku": "b",
      },
    ],
  },
  {
    "accumulate_pattern": 2,
    "items": [
      {
        "order_id": 2,
        "sku": "c",
      },
    ],
  },
]
`;
